force-app/main/default/classes/DMLOperations.cls

```diff
@@ -181,14 +256,64 @@ public with sharing class DMLOperations {
 	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
 		// Loop through each Contact
 
-		// Extract the Account Name from Contact's LastName
-
-		// Get the account based on the name or create one if it doesn't exist
-
-		// Link the Contact to its Account using the Account Id
-
-		// Upsert the Contacts
-	}
+        //i was unable to do this question myself
+
+            if (contactsToUpsert.isEmpty()) {
+                System.debug('Error: List is empty');
+                return; // Exit the method if the list is empty
+            }
+        
+            // Extract the last names from the contacts
+            Set<String> contLastNameSet = new Set<String>();
+            for (Contact cont : contactsToUpsert) {
+                contLastNameSet.add(cont.LastName);
+            }
+        
+            // Query existing Accounts with names that match the contact last names
+            List<Account> existingAccounts = [SELECT Id, Name FROM Account WHERE Name IN :contLastNameSet];
+        
+            // Map to hold the existing account names for quick lookup
+            Map<String, Account> accountMap = new Map<String, Account>();
+            for (Account acc : existingAccounts) {
+                accountMap.put(acc.Name, acc);
+            }
+        
+            // List to hold new accounts to be created
+            List<Account> accountsToInsert = new List<Account>();
+        
+            // Loop through the contact list to ensure each has a corresponding account
+            for (Contact cont : contactsToUpsert) {
+                if (!accountMap.containsKey(cont.LastName)) {
+                    // Account does not exist, create a new one
+                    Account newAcc = new Account(Name = cont.LastName);
+                    accountsToInsert.add(newAcc);
+                    accountMap.put(cont.LastName, newAcc); // Add to map to link later
+                }
+            }
+        
+            // Insert the new accounts
+            if (!accountsToInsert.isEmpty()) {
+                insert accountsToInsert;
+        
+                // Update the account map with the newly inserted accounts
+                for (Account acc : accountsToInsert) {
+                    accountMap.put(acc.Name, acc);
+                }
+            }
+        
+            // Link the contacts to the newly created or existing accounts
+            for (Contact cont : contactsToUpsert) {
+                if (accountMap.containsKey(cont.LastName)) {
+                    cont.AccountId = accountMap.get(cont.LastName).Id;
+                }
+            }
+        
+            // Upsert the contacts with the associated account IDs
+            upsert contactsToUpsert;
```

Comment:
Wonderful job structuring this logic to ensure you didn't put any DML or queries inside your loops! This isn't an easy feat!